### ISSUE TITLE
Can't get Arma 3 server

### ISSUE THREAD

**Issue by Gerardc9** [Austinb/GameQ#420]
Since some time ago, can't connect to Arma 3 server. 

I'm getting all time "Impossible to connect with the server", any idea about this? I've tested with the V3 branch. 

EDIT: I got  back this data if I do DD

`Collection {#591 ▼
  #items: array:19 [▼
    "gq_address" => "151.xx.xx.xx"
    "gq_dedicated" => null
    "gq_gametype" => null
    "gq_hostname" => null
    "gq_joinlink" => "steam://connect/151.xx.xx.xx:2302/"
    "gq_mapname" => null
    "gq_maxplayers" => null
    "gq_mod" => null
    "gq_name" => "Arma3"
    "gq_numplayers" => null
    "gq_online" => false
    "gq_password" => null
    "gq_port_client" => 2302
    "gq_port_query" => 2303
    "gq_protocol" => "source"
    "gq_transport" => "udp"
    "gq_type" => "arma3"
    "players" => []
    "teams" => []
  ]
}`

Any idea @Austinb ? Many thanks 👍

**Comment by ALTITUDE-DEV-FR** [comment]
Same please help !  @Austinb

**Comment by Austinb** [comment]
Make sure you have the correct IP and client port.  Also make sure your host you are running the script on will allow you to make outside queries.  Many hosting services block outgoing requests for queries to prevent abuse (as part of a larger firewall ruleset). 

Please provide the ip:port (in private message if you dont want it public) so I can test and see if the ports are correct.

**Comment by Gerardc9** [comment]
Hi @Austinb, the server that I'm testing is the following: 151.80.47.174:2302. Regards!

**Comment by Austinb** [comment]
It appears something has changed in the response.  The parsing is failing when trying to parse the mods section of the rules since Arma3 uses its own protocol for the rules.  No matter what server I pull from game tracker it always says it has 9 mods when they do not. This causes the script to run out of response data to parse and it throws and Exception.  You can see the exception if you enable debugging. 

@sifex Can you provide any assistance?  There is an empty string that readInt8() returns 9 which is where it gets the number of mods in the response.

**Comment by sifex** [comment]
I'll have a look into it tomorrow (hopefully)

**Comment by Gasjki** [comment]
Any update ?

EDIT: I'll start an investigation too. Please keep us up-to-date too and I'll do the same.

**Comment by Gasjki** [comment]
Hi @Austinb & @sifex ,
I made few tests on local and I discovered that after this line: https://github.com/Austinb/GameQ/blob/v3/src/GameQ/Query/Native.php#L114 everything is still fine, but the problem is on next `stream_select()`. So, for my test, I tried to stop execution right after `stream_set_blocking()` and I did this (following this script - https://oxidemod.org/threads/php-simple-server-query.2374/): 

            // Set blocking mode.
            stream_set_blocking($this->socket, TRUE /**$this->blocking**/);

            $this->writeSocket("\xFF\xFF\xFF\xFF\x54Source Engine Query\x00");
            $response = fread($this->socket, 8196);
            @fclose($this->socket);

            $packet = explode("\x00", substr($response, 6), 5);

            $server['name']        = $packet[0];
            $server['map']         = $packet[1];
            $server['game']        = $packet[2];
            $server['description'] = $packet[3];
            $inner                 = $packet[4];
            $server['players']     = ord(substr($inner, 2, 1));
            $server['playersmax']  = ord(substr($inner, 3, 1));
            $server['password']    = ord(substr($inner, 7, 1));
            $server['vac']         = ord(substr($inner, 8, 1));

            echo '<pre>';
            print_r($server);
            echo '</pre>';
            die;
And surprise! I have a result!!! 

```
Array
(
    [name] => [GER] Zero-One Altis Life | TS3: ts.zero-one.cc
    [map] => Altis
    [game] => Arma3
    [description] => Zero-One Altis Life | www.zero-one.cc
    [players] => 155
    [playersmax] => 230
    [password] => 0
    [vac] => 0
)
```

IP & port: 149.202.223.195:2302

Also, you MUST set `stream_set_blocking()` as TRUE, otherwise, array is empty.
Hope this will help you guys!

**Comment by Austinb** [comment]
The initial issue is in the response from a change in the protocol from Arma's side breaking the rules response parsing.  I will try to take a look again and see if they have made any public comments about the change.  I just tested your server and now I am getting a bad response at the Source protocol level which leads me to believe they changed more now,  You can see the errors if you set debug to true `$gq->setOption('debug', true);` when you make your query.  By default if the protocol fails at any level it just returns as if the server is offline.  With debug it will throw the exception so you can see the issue. 

@sifex Did you ever get a chance to look at the change or find any documentation for it?  I can make it work if I knew what the change was to the protocol.

**Comment by Austinb** [comment]
@Gasjki Just for info your server is returning tow packets with the response code "A" which is not on the list of code from the Source protocol (https://developer.valvesoftware.com/wiki/Server_queries#Source_Server). The response type is the character after the first 4 header bytes `<FF>`. The third entry with "I" is the server details info (A2S_INFO) in the Source docs which is where the server name, map and other info is housed. 

```sh
Array
(
    [0] => <FF><FF><FF><FF>A<CA><C8>S<AE>
    [1] => <FF><FF><FF><FF>A<CA><C8>S<AE>
    [2] => <FF><FF><FF><FF>I^Q[GER] Zero-One Altis Life | TS3: ts.zero-one.cc^@Altis^@Arma3^@Zero-One Altis Life | www.zero-one.cc^@^@^@<E0><E6>^@dw^@^@1.82.144647^@<B1><FE>^H^A,K<F5><9D>&@^Abt,r182,n0,s7,i3,mf,lf,vt,dt,trpg,g65545,hca42103d,f0,c-2147483648--2147483648,pw,e15,j0,k0,^@<92><A3>^A^@^@^@^@^@
)
```

I pulled some servers at random from gametracker and found a few with the same issue that had Altis running but also some with that same mod that worked fine.  I am not sure why those calls are not being returned properly from the server's side.  It did the same thing with or without the blocking enabled on the socket.  

The servers that returned properly still had the length issue due to a change somewhere in the rules response.  I am going to just drop the rules for now so the protocol will work without that part of the response.

**Comment by Austinb** [comment]
Pushed change to drop rules parsing into master.  You will need to change your composer.json to use the master branch if you want to use the change.  Will NOT push this change into a release so it will need to be fixed before next release.

**Comment by Gasjki** [comment]
I tried to enabled debug on your script and guess what? I don't get any error... Excepting debug, what I'm supposed to do enable/disable in order to get those errors ?

**Comment by Austinb** [comment]
I would pull down the v3 master code and run it normally removing any changed you made.  The commit I made last week removes the rules issue but the "A" response issue is separate from this one and I am not sure what causes it.   If you enable debug and that "A" response comes back it will throw an exception to your console/page.  

With the change above you should not see any more errors due to the changed response for the rules. I have not had time to look at this again yet to see what is going on with the response.

**Comment by Austinb** [comment]
Has there been any changes to the protocol or the documentation to take into account this breaking change in the response?  Has it by chance self-corrected?

**Comment by Gerardc9** [comment]
Hi @Austinb, I really don't know exactly if has been changed something about the protocols of Arma 3. And nope, has not be self-corrected :(

**Comment by Austinb** [comment]
What are the DLCs that are available for use?  All of the servers are responding with the same mod count which is wrong which makes me think something in the response has changed due to more dlcs or other changes in that grouping.

**Comment by Austinb** [comment]
Ok so from what I can tell it appears the other DLCs have been released but the dlc bit has not been updated to reflect the change.  When querying servers I am getting a 119 which is a sum power of 2 for 7 possible entries (including 0).  However I see there are 8 DLCs listed and all of them are available.

**Comment by Austinb** [comment]
After a bit of testing it appears that the dlc flag is the issue but it is due to improper handling of the values.  This should be done in the same way the 5 set is done with offsets 0 thru 7 as on/off instead of using power of 2.  This will change how the code is run.  Working on an update and hope to have something today to get this done.  This was an oversight when they first changed the query breaking from the A2S standard.

**Comment by Austinb** [comment]
So I have finally figured out the issue.  The DLC bit appears to be wrong or not being updated.  When either the bit flag 0-7 or power of 2 the value being returned is wrong.   

Example:
For the server 195.249.41.71:2302 the DLC byte value is 119 which is only 6 of the 8 DLCs enabled.  However there are signatures in the response for all 8 DLCs.  If I manually change the DLCs to be 8 (DLC flag 255) it processes the extra 8 bytes of DLC info (2 @ 4 bytes each) and the response works as expected. Also using the 119 and burning the 8 bytes at the end of that section works the same way.

Does anyone have a contact to their developers?

@Memphis017 You have helped with this in the past.  So you can see what my response looks like this is what I have but I am guessing on which mod is what. It appears https://community.bistudio.com/wiki/Arma_3_ServerBrowserProtocol2 needs to be updated since there are now 8 dlcs and all 8 bits in that DLC entry should be enabled.

For a value of 119 the flags are 01110111 (as per the docs 7 to 0, left to right) when it should be 255 (11111111). 

I am going to hold on to these changes until there is a fix by a patch or I can find some way to code around it.  For now the rules will have to remain empty which kind of sucks.

**Comment by Memphis017** [comment]
@Austinb  

Indeed and as usual they haven't updated the doc... My guess is that they now use the previously reserved bit after the dlc one to extend it. If you look at this bit you will find your two missing bits (3 for my server), so you need to add up these two to have the correct value of the DLC bit.

#433

**Comment by Austinb** [comment]
Fixed in PR #438. Open again if any issues.
